### PR TITLE
Unify the input order for text (NLG) metrics - BLEU, SacreBLEU, TER, CHRF

### LINKED ISSUE #686: Unify the input order for `text` (NLG) metrics
## 🚀 Enhancement

Unify the input order for `text` (NLG) metrics

### Motivation

We aim to unify text (NLG) metrics in a way that all of them will allow evaluating a hypothesis against multiple references. Nonetheless, there remains another distinction as some metrics presume `preds, targets` input order, while some expect `targets, preds` order. With respect to the whole TM ecosystem, we will go with `preds, targets`.

### Pitch

Change input order for the following NLG metrics:
- `BLEUScore` - B.C (add a warning suggested [here](https://github.com/PyTorchLightning/metrics/pull/687#issuecomment-999781877))
- `CHRFScore` - non B.C. (a new metrics in `v0.7`)
- `SacreBLEUScore` - B.C (add a warning suggested [here](https://github.com/PyTorchLightning/metrics/pull/687#issuecomment-999781877))
- `TER` - non B.C. (a new metrics in `v0.7`)

Standardize naming convention to have: `predictions` and `targets` everywhere.

### Alternatives

Keep it as it is.

@SkafteNicki @Borda @ashutoshml

### PR DESCRIPTION
**Pull request by ashutoshml**
## What does this PR do?

Fixes #686 

## Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃
Yes.

### CODE REVIEW

File: tests/text/test_bleu.py [hunk 1]
```diff
@@ -27,12 +27,12 @@
 smooth_func = SmoothingFunction().method2
 
 
-def _compute_bleu_metric_nltk(list_of_references, hypotheses, weights, smoothing_function, **kwargs):
-    hypotheses_ = [hypothesis.split() for hypothesis in hypotheses]
-    list_of_references_ = [[line.split() for line in ref] for ref in list_of_references]
+def _compute_bleu_metric_nltk(predictions, list_of_targets, weights, smoothing_function, **kwargs):
```
Comment (Borda) [inline tests/text/test_bleu.py:30]:
```suggestion
def _compute_bleu_metric_nltk(preds, targets, weights, smoothing_function, **kwargs):
```

File: tests/text/test_bleu.py [hunk 2]
```diff
@@ -49,74 +49,74 @@ def _compute_bleu_metric_nltk(list_of_references, hypotheses, weights, smoothing
...
+            input_order=INPUT_ORDER.PREDS_FIRST,
         )
 
 
 def test_bleu_empty_functional():
-    hyp = [[]]
-    ref = [[[]]]
-    assert bleu_score(ref, hyp) == tensor(0.0)
+    prediction = [[]]
+    target = [[[]]]
```
Comment (Borda) [inline tests/text/test_bleu.py:101]:
```suggestion
    predictions = [[]]
    targets = [[[]]]
```

**Review by Borda — COMMENTED** [review]
lets use everywhere `preds` and `targets`

File: tests/text/test_bleu.py (hunk 1 above)
Comment (ashutoshml) [reply tests/text/test_bleu.py:30]: Ok. I will update by tomorrow/day-after.

Comment (ashutoshml) [reply tests/text/test_bleu.py:30]: @Borda @stancld: The requirements of the issue are being changed every iteration. Let us finalize one. If `preds` and `targets` are final, please update the requirements of issue #686. I will modify those 4 metrics, accordingly.

Comment (stancld) [reply tests/text/test_bleu.py:30]: @ashutoshml It looks like the whole `text` package has a different naming than the rest of `tm` package, so this was likely led to our confusion.

cc: @Borda (shouldn't we have `preds, target`, without `s` in `targets` as it is the case in other packages within `tm`?

Comment (Borda) [reply tests/text/test_bleu.py:30]: my typo, primary is consistency across all TM... :)

**Comment by Borda** [comment]
@ashutoshml how is it going here, what is missing?
@stancld could you assist to get it to merge quickly...

**Comment by stancld** [comment]
> @ashutoshml how is it going here, what is missing? @stancld could you assist to get it to merge quickly...

@Borda I think we were not fully sure with the naming. But @ashutoshml, it's `preds, target` everywhere for sure, so ping me whenever this PR is ready for review :] Sorry for some confusion

**Comment by Borda** [comment]
Let's keep it moving, can we limit this PR just to switching the order as it is stated in the PR title and do the rename package-wide in another PR :rabbit: @stancld @ashutoshml

**Comment by ashutoshml** [comment]
> Let's keep it moving, can we limit this PR just to switching the order as it is stated in the PR title and do the rename package-wide in another PR as it will require another deprecation/compatibility edits 🐰 @stancld @ashutoshml cc: @justusschock @SkafteNicki

If the naming convention is not an issue, then the PR is ready for review. The order issue has been fixed.

**Comment by Borda** [comment]
> If the naming convention is not an issue, then the PR is ready for review. The order issue has been fixed.

yes, but in such a case pls ever all name changes or you need to also include all deprecation for argument's name changes... :]

**Comment by ashutoshml** [comment]
> > If the naming convention is not an issue, then the PR is ready for review. The order issue has been fixed.
> 
> yes, but in such a case pls ever all name changes or you need to also include all deprecation for argument's name changes... :]

I see. I'll revert all naming to previous ones then, and only modify the order. I will make the changes latest by tomorrow.

**Comment by ashutoshml** [comment]
@Borda @stancld Please have a look

File: torchmetrics/functional/text/bleu.py [hunk 3]
```diff
@@ -174,6 +180,10 @@ def bleu_score(
         [2] Automatic Evaluation of Machine Translation Quality Using Longest Common Subsequence
         and Skip-Bigram Statistics by Chin-Yew Lin and Franz Josef Och `Machine Translation Evolution`_
     """
+    warnings.warn(
+        "Input order of targets and preds were changed to predictions firsts and targets \
+                   second in v0.7. Warning will be removed in v0.8"
```
Comment (Borda) [inline torchmetrics/functional/text/bleu.py:185]:
```suggestion
        "Input order of targets and preds were changed to predictions firsts and targets second in v0.7."
        " Warning will be removed in v0.8."
```

File: torchmetrics/functional/text/sacre_bleu.py [hunk 4]
```diff
@@ -321,6 +322,10 @@ def sacre_bleu_score(
         [3] Automatic Evaluation of Machine Translation Quality Using Longest Common Subsequence
         and Skip-Bigram Statistics by Chin-Yew Lin and Franz Josef Och `Machine Translation Evolution`_
     """
+    warnings.warn(
+        "Input order of targets and preds were changed to predictions firsts and targets \
+                   second in v0.7. Warning will be removed in v0.8"
```
Comment (Borda) [inline torchmetrics/functional/text/sacre_bleu.py:327]:
```suggestion
        "Input order of targets and preds were changed to predictions firsts and targets second in v0.7."
        " Warning will be removed in v0.8."
```

File: torchmetrics/text/bleu.py [hunk 5]
```diff
@@ -81,7 +82,10 @@ def __init__(
             process_group=process_group,
             dist_sync_fn=dist_sync_fn,
         )
-
+        warnings.warn(
+            "Input order of targets and preds were changed to predictions firsts and targets \
+                    second in v0.7. Warning will be removed in v0.8"
```
Comment (Borda) [inline torchmetrics/text/bleu.py:87]:
```suggestion
            "Input order of targets and preds were changed to predictions firsts and targets second in v0.7."
            " Warning will be removed in v0.8."
```

**Review by Borda — APPROVED** [review]

**Comment by Borda** [comment]
> @Borda @stancld Please have a look

cool, @ashutoshml do you want to prepare the next one with rename arguments to be TM-wide consistent or @stancld?

**Review by stancld — APPROVED** [review]
LGTM! Nice work, thanks a lot @ashutoshml

**Comment by ashutoshml** [comment]
> > @Borda @stancld Please have a look
> 
> cool, @ashutoshml do you want to prepare the next one with rename arguments to be TM-wide consistent or @stancld?

That might need changing all the files. Would it be ok to split the metrics into multiple PRs? 4-4-4 or 6-6 since there are 12 text metrics.

**Comment by Borda** [comment]
> That might need changing all the files. Would it be ok to split the metrics into multiple PRs? 4-4-4 or 6-6 since there are 12 text metrics.

sure, that sounds good to me... if you split it 6-6 you can also divide work so 6 for @ashutoshml and 6 for @stancld :rabbit:

**Comment by stancld** [comment]
> > That might need changing all the files. Would it be ok to split the metrics into multiple PRs? 4-4-4 or 6-6 since there are 12 text metrics.
> 
> sure, that sounds good to me... if you split it 6-6 you can also divide work so 6 for @ashutoshml and 6 for @stancld? 🐰

@ashutoshml Just let me know which metrics you are taking and I'll handle the remaining ones :]

**Comment by stancld** [comment]
@Borda 

Those failing tests seem to be unrelated to this PR:

```
ERROR torchmetrics/audio/pesq.py - ValueError: numpy.ndarray size changed, ma...
ERROR torchmetrics/functional/audio/pesq.py - ValueError: numpy.ndarray size ...
ERROR tests/audio/test_pesq.py - ValueError: numpy.ndarray size changed, may ...
ERROR tests/audio/test_pesq.py - ValueError: numpy.ndarray size changed, may ...
```

Can it be merged then? :]

**Comment by Borda** [comment]
> Those failing tests seem to be unrelated to this PR:

do they fail also on master? :]

**Comment by stancld** [comment]
> > Those failing tests seem to be unrelated to this PR:
> 
> do they fail also on master? :]

With the most recent commit #705, these tests are currently failing on the `master` too. I can have a look at them

**Comment by ashutoshml** [comment]
> > > That might need changing all the files. Would it be ok to split the metrics into multiple PRs? 4-4-4 or 6-6 since there are 12 text metrics.
> > 
> > 
> > sure, that sounds good to me... if you split it 6-6 you can also divide work so 6 for @ashutoshml and 6 for @stancld? 🐰
> 
> @ashutoshml Just let me know which metrics you are taking and I'll handle the remaining ones :]

@stancld I'll take `cer, ter, wer, sacrebleu, rouge, squad`. Kindly let me know when you create an issue and assign it to me.

**Review by awaelchli — APPROVED** [review]

**Pull request merged**